### DISCUSSION TITLE
What is the time frame for support lifetime, enum, GPU

### DISCUSSION THREAD

**Discussion by Eprahim-taha** [modular/modular#1536] [General]
⚠️ This article has been translated and may contain some misspellings 

I am working on creating an open source library for machine vision in Mojo which will provide a beautiful user interface and give the user high productivity capability, for example, the ability to use machine vision algorithms with different types of Arduino and avoid the C++ nightmare, but the challenge I have now is the limited capabilities of  Mojo.  As the language does not support the GPU, which is very important when talking about machine vision.  Lacks the ability to return references, and I don't know the time frame in which I can safely return a reference in Mojo and safely store references in structures

**Comment by rarebreed** [comment]
According to the 2023 keynote, when the SDK for the Max Engine is released, it should have support for GPU.

What I don't know is whether MAX Engine is only available as a cloud service, or if it can be run and tested locally too (I might have missed it somewhere).   But they said should be available first quarter of 2024.  Although I hope by that time, they have also released lifetime/references and Algebraic Data Types, and won't initially release it with mojo 0.6.0

https://www.modular.com/blog/key-announcements-from-modcon-2023

**Reply by Eprahim-taha** [reply]
They haven't released it yet enum, Smart pointers, lifetime
